pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go

```diff
@@ -132,6 +136,79 @@ func quantizeTS(ts hlc.Timestamp, granularity time.Duration) hlc.Timestamp {
 	}
 }
 
+func (p *rangefeed) handleRangefeedEvent(ctx context.Context, e *kvpb.RangeFeedEvent) error {
+	switch t := e.GetValue().(type) {
+	case *kvpb.RangeFeedValue:
+		if p.cfg.Knobs.OnRangeFeedValue != nil {
+			if err := p.cfg.Knobs.OnRangeFeedValue(); err != nil {
+				return err
+			}
+		}
+		stop := p.st.RangefeedBufferValue.Start()
+		if err := p.memBuf.Add(
+			ctx, kvevent.MakeKVEvent(e),
+		); err != nil {
+			return err
+		}
+		stop()
+	case *kvpb.RangeFeedCheckpoint:
+		ev := e.ShallowCopy()
+		ev.Checkpoint.ResolvedTS = quantizeTS(ev.Checkpoint.ResolvedTS, p.cfg.WithFrontierQuantize)
+		if resolvedTs := ev.Checkpoint.ResolvedTS; !resolvedTs.IsEmpty() && resolvedTs.Less(p.cfg.Frontier) {
+			// RangeFeed happily forwards any closed timestamps it receives as
+			// soon as there are no outstanding intents under them.
+			// Changefeeds don't care about these at all, so throw them out.
+			return nil
+		}
+		if p.knobs.ShouldSkipCheckpoint != nil && p.knobs.ShouldSkipCheckpoint(t) {
+			return nil
+		}
+		stop := p.st.RangefeedBufferCheckpoint.Start()
+		if err := p.memBuf.Add(
+			ctx, kvevent.MakeResolvedEvent(ev, jobspb.ResolvedSpan_NONE),
+		); err != nil {
+			return err
+		}
+		stop()
+	case *kvpb.RangeFeedSSTable:
+		// For now, we just error on SST ingestion, since we currently don't
+		// expect SST ingestion into spans with active changefeeds.
+		return errors.Errorf("unexpected SST ingestion: %v", t)
+	case *kvpb.RangeFeedBulkEvents:
+		// TODO(#138346): We can handle these more gracefully once we
+		// migrate to the new rangefeed client. Until then, this is
+		// still an improvement over not using these.
+		for _, e := range t.Events {
```

Comment:
Should we maybe add a timer metric for this event type too? Although we'd end up duplicating measurements for the inner event types (unless we added a flag or something to the func to signify whether it's part of a bulk event), but maybe that's desirable so that we can still see the distribution? 🤷‍♂️

Reply:
dont think that's necessary. the main purpose of that timer is to observe buffer adds. that shouldnt depend on whether its coming from a bulk event or a regular event